Add ability to grant access to arrest warrants

halworsen opened the pull request:
:cl:Atebite
rscadd: Anyone with access to change IDs can now authorize access on arrest warrants. Holoprojectors can then be used as ID cards as long as the selected warrant is active. Warrants copy all the job access of the job listed in the person's crew records, except for command access
tweak: Changing most information in crew records now requires ID changing accesss
/:cl:

CL and title should be pretty descriptive.

PsiOmegaDelta commented on code/_helpers/mobs.dm:
```diff
@@ -265,4 +265,11 @@ proc/age2agedescription(age)
+	for(var/mob/living/carbon/human/H in GLOB.human_mob_list)
+		if(H.real_name == name)
+			return H
+	return 0
```
Don't mix return types. 
`null`or not returning anything at all is appropriate here.

PsiOmegaDelta reviewed (commented):
This also doesn't handle the situation where someone has had a change of job (or had access changes in general) after joining the round.

halworsen commented:
I don't think it's actually possible to track job/access changes on the mob, since those are tied to ID cards (which themselves don't have any link to their owners except for their name) and crew records.

I could change crew records so that changing job also requires ID changing access, though.

halworsen commented:
Changed it to rely 100% on crew records. If the name you put out a warrant for is in the crew records, it'll look up the job listed in their records and use access from that job. To prevent abuse, I've changed crew records so that most entries can only be changed with ID changing access.

halworsen committed:
Add ability to grant access to arrest warrants

SparklySheep commented:
Why even have ID's when arrest warrants are just ID+?

halworsen commented:
Seeing a lot of "confused" on this. Let me clarify:

You need ID changing access (read: only the XO/CO) to authorize access on a warrant. When you authorize access on a warrant, it takes the name on the warrant and looks it up in the crew records. If it finds someone in crew records with that name, it looks at their listed job and adds that job's access to the warrant. For example, if you put out a warrant on Joe Schmoe, who is listed in the crew records as a chemist, the warrant gets all the access that a chemist has by default when you authorize access.

To actually make use of that access, you need a warrant projector. When you select a warrant on your projector, you can use the projector as an ID. The access you have on your projector depends on the access on the warrant you've selected. So in the case of Joe Schmoe, using his warrant would make your holoprojector work as a chemist's ID *so long as the warrant is active*. It doesn't change your ID card, but when held in the hand it can work together with your ID to grant you additional access.

Another important thing to note is that this can never give a warrant command access. So you can't get access to the bridge, teleporters, captain's office, etc. with this. This is also only available for arrest warrants. Search warrants can never give you any access.

halworsen committed:
Remove now-unused helper proc

PsiOmegaDelta commented:
How does this handle multiple crew members with the same name?

halworsen committed:
Lookup job too when authorizing access on warrants

halworsen commented:
Oopsie. Changed it to take name *and* job for arrest warrants. The name/job combo has to match the crew records, so you can't just enter a valid name in the crew records and then any job you want. They need to actually be listed as the job you enter.

In the unlikely case that someone has the same name & job it obviously doesn't matter since the outcome is the same.

NullSnapshot commented:
I have to say, this is actually a pretty clever way to handle the access issue when conducting warrant arrests. I wouldn't have thought of it.

Legodude9 commented:
I disagree with this, as a Warrent isn't supposed to allow you to self autonomously walk into some place, if its an arrest warrent, you need to check they will comply, same with search warrents. Also gives security a massive power boost, which it shouldn't have from a balance side. Also doesn't take into acount that the Deck Officer and possibly some other postions outside of security have Warrent Projectors.

afterthought2 commented:
As is explained above, you need the access modification access to enable this feature on the warrant. It also needs a valid crew record to work. Doesn't seem too imbalanced.

comma commented:
Yeah it's same as XO printing out buncha guest passes and giving them to officers, sans painful UI juggle.
